dopamine/utils/threading_utils.py

```diff
@@ -0,0 +1,158 @@
+
+  For each specified attribute, we create a global default value that is
+  object-specific and not thread-specific. This gloal deafault value is stored
+  in an internal attribute that is named `_attr_default` where `attr` is the
+  name of the specified attribute.
+
+  Args:
+    name: str, name of the exposed attribute.
+  Returns:
+    str, name of the internal attribute storing thread local value.
+  """
+  return '_' + name + '_default'
+
+
+def _add_property(cls, attr_name):
+  """Adds a property to a given class.
+
+  The a setter, getter and deleter are added to the given class with to the
+  provided attribute name.
+  These methods actually apply to an intenal variable that is thread-specific.
+  Hence the result of these methods depends on the local thread.
+
+  Note that when the getter is called and the local attribute is not found the
+  getter will initialize the local value to the global default value.
+  See `initialize_local_attributes` for more details.
+
+  Args:
+    cls: A class to add the poperty to.
+    attr_name: str, name of the property to create.
+  """
+  def _set(self, val):
+    setattr(self, _get_internal_name(attr_name), val)
+
+  def _get(self):
+    if not hasattr(self, _get_internal_name(attr_name)):
+      if not hasattr(self, _get_default_value_name(attr_name)):
+        raise AttributeError(
+            'Local value for attribute `{}` has not been set. You can '
+            'initialize it locally (`self.{} = <initial-value>`) or set a '
+            'global value using the `initialize_local_attributes` '
+            'helper.'.format(attr_name, attr_name))
+      _default_attr_fn = getattr(self, _get_default_value_name(attr_name))
+      if not callable(_default_attr_fn):
+        raise AttributeError('Default value initializer must be callable.')
+      _set(self, _default_attr_fn())
+    return getattr(self, _get_internal_name(attr_name))
+
+  def _del(self):
+    delattr(self, _get_internal_name(attr_name))
+
+  setattr(cls, attr_name, property(_get, _set, _del))
+
+
+def local_attributes(attributes):
+  """Creates a decorator that add properties to the decorated class.
+
+  Args:
+    attributes: List[str], names of the wrapped attributes to add to the class.
+  Returns:
+    A class decorator.
+  """
+  def _decorator(cls):
+    for attr_name in attributes:
+      _add_property(cls, attr_name)
+    return cls
+  return _decorator
+
+
+def initialize_local_attributes(obj, **kwargs):
+  """Sets global default values for local attributes.
+
+  Each attribute has a global default value initializer and local values that
+  are specific to each thread.
+  In each thread, the first time the getter is called it is initialized by
+  calling the global default value initializer. This helper function is to set
+  these default value initializers.
+
+  Args:
+    obj: The object that has the local attributes.
+    **kwargs: The default value for each local attribute.
```

Comment:
With the new interface (which I like, BTW), this part of the docstring is no longer 100% precise.

Reply:
Good catch - replaced to `The default callable initializer for each local attribute.`